packages/effect/src/config/ConfigProvider.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * @since 4.0.0
+ */
+import * as Effect from "../Effect.js"
+import { constant, dual } from "../Function.js"
+import { PipeInspectableProto } from "../internal/core.js"
+import type { Pipeable } from "../Pipeable.js"
+import * as ServiceMap from "../ServiceMap.js"
+import * as String from "../String.js"
+import { type ConfigError, filterMissingData, MissingData } from "./ConfigError.js"
+
+/**
+ * @since 4.0.0
+ * @category Models
+ */
+export interface ConfigProvider extends Pipeable {
+  readonly currentPath: ReadonlyArray<string>
+  readonly append: (path: string) => ConfigProvider
+  readonly withValue: (value: string) => ConfigProvider
+
+  readonly load: Effect.Effect<string, ConfigError>
+  readonly loadFromPath: (path: ReadonlyArray<string>) => Effect.Effect<string, ConfigError>
+}
+
+/**
+ * @since 4.0.0
+ * @category References
+ */
+export class CurrentConfigProvider extends ServiceMap.Reference("effect/config/ConfigProvider/CurrentConfigProvider", {
+  defaultValue: (): ConfigProvider => fromEnv()
+}) {}
+
+/**
+ * @since 4.0.0
+ * @category Constructors
+ */
+export const make = (load: (path: ReadonlyArray<string>) => Effect.Effect<string, ConfigError>): ConfigProvider =>
+  makeProto({
+    currentPath: [],
+    loadFromPath: load
+  })
+
+const makeProto = (options: {
+  readonly currentPath: ReadonlyArray<string>
+  readonly loadFromPath: (path: ReadonlyArray<string>) => Effect.Effect<string, ConfigError>
+  readonly append?: (path: string) => ConfigProvider
+}): ConfigProvider => {
+  const self = Object.create(Proto)
+  self.currentPath = options.currentPath
+  self.loadFromPath = options.loadFromPath
+  if (options.append) {
+    self.append = options.append
+  }
+  return self
+}
+
+const Proto = {
+  ...PipeInspectableProto,
+  toJSON(this: ConfigProvider) {
+    return {
+      _id: "ConfigProvider",
+      currentPath: this.currentPath
+    }
+  },
+  append(this: ConfigProvider, path: string): ConfigProvider {
+    return makeProto({
+      ...this,
+      currentPath: [...this.currentPath, path]
+    })
+  },
+  withValue(this: ConfigProvider, value: string): ConfigProvider {
+    return makeProto({
+      ...this,
+      loadFromPath: constant(Effect.succeed(value))
+    })
+  },
+  get load(): Effect.Effect<string, ConfigError> {
+    return (this as any).loadFromPath((this as any).currentPath)
+  }
+}
+
+/**
+ * @since 4.0.0
+ * @category Constructors
+ */
+export const fromEnv = (options?: {
+  readonly pathDelimiter?: string | undefined
+  readonly environment?: Record<string, string | undefined> | undefined
+}): ConfigProvider => {
+  const env = options?.environment ?? {
+    ...process.env,
+    ...(import.meta as any).env
```

Comment:
Won't these fail if the globals don't exist?

Reply:
🤦‍♂️ yup